### PR TITLE
Response headers class

### PR DESCRIPTION
**Pull request by poratuk**

### CODE REVIEW

File: src/ResponseHeaders.php [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+     *
+     * @var bool
+     */
+    protected $isStale;
+
+    /**
+     * Create header array from resived array in the Header дшые
+     *
+     * @param array $incomingArray
+     * @response Header class object
```
Comment (jasny) [inline src/ResponseHeaders.php]: `@response`?? no

File: src/ResponseHeaders.php [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+     */
+    public function __construct($incomingArray = [])
```
Comment (jasny) [inline src/ResponseHeaders.php]: The constructor shouldn't take any arguments. Just creating a `ResposneHeaders` object, shouldn't change the global state.

File: src/ResponseHeaders.php [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+ */
+class ResponseHeaders extends Headers
+{
+    
+    /**
+     * HTTP headers
+     *
+     * @var bool
+     */
+    protected $isStale;
```
Comment (jasny) [inline src/ResponseHeaders.php]: Should be only `true` or `false`

File: src/ResponseHeaders.php [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    }
+    
+    /**
+     * For the current object simple chek  if current object 
+     * can be used for set or modify or remove headers
+     *
+     * @throws \InvalidArgumentException
+     */
+    protected function assertStale()
```
Comment (jasny) [inline src/ResponseHeaders.php]: Function should be named `assertNotStale` since that's what's you're asserting.

File: src/ResponseHeaders.php [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    {
+        if ($this->isStale !== null) {
+            throw new \InvalidArgumentException("Can not change stale object");
+        }
+    }
+
+    /**
+     * Change current object to stale, move all headers to the variable
+     */
+    protected function setClassStale()
```
Comment (jasny) [inline src/ResponseHeaders.php]: Rename to `turnStale`, so it's the same as in [`ServerRequest\GlobalEnvironment`](https://github.com/jasny/http-message/blob/master/src/ServerRequest/GlobalEnvironment.php#L95).

File: src/ResponseHeaders.php [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+    /**
+     * Retrieves all message header values.
+     *
+     * The keys represent the header name as it will be sent over the wire, and
+     * each value is an array of strings associated with the header.
+     *
+     * // Represent the headers as a string
+     * foreach ($message->getHeaders() as $name => $values) {
+     * echo $name . ': ' . implode(', ', $values);
```
Comment (jasny) [inline src/ResponseHeaders.php]: Do correct indenting, also in comments

File: src/ResponseHeaders.php [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            return parent::hasHeader($name);
+        }
+        
+        $headers = headers_list();
+        foreach ($headers as $header) {
+            list($key) = explode(': ', $header);
+            if (strtolower($key) == strtolower($name)) {
+                return true;
+            }
+        }
```
Comment (jasny) [inline src/ResponseHeaders.php]: This code is duplicated (twice). Instead add `protected function getRawHeader()`, that returns the header value.

File: src/ResponseHeaders.php [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    {
+        $this->isStale = true;
+        
+        $list = headers_list();
+        foreach ($list as $header) {
+            list($key, $value) = explode(': ', $header);
+            $this->headers[strtolower($key)] = ['name' => $key, 'values' => explode(', ', $value)];
+        }
+    }
+
```
Comment (jasny) [inline src/ResponseHeaders.php:52]: Also add an `isStale` method. This will help in the tests.

File: tests/ResponseHeadersTest.php [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        $headers = $this->headers->getHeaders();
+        $this->assertSame([], $headers);
+    }
+
+    public function testWithHeader()
+    {
+        $header = $this->headers->withHeader('Foo-Zoo', 'red & blue');
+        
+        $this->assertInstanceof(ResponseHeaders::class, $header);
+        $this->assertNotSame($this->headers, $header);
```
Comment (jasny) [inline tests/ResponseHeadersTest.php:37]: Test that `$this->headers` is stale.

File: tests/ResponseHeadersTest.php [hunk 10]
```diff
@@ -0,0 +1,271 @@
+<?php
+
+namespace Jasny\HttpMessage;
+
+use Jasny\HttpMessage\Tests\AssertLastError;
+use PHPUnit_Framework_TestCase;
+
+/**
+ * @covers Jasny\HttpMessage\ResponseHeaders
+ * @runTestsInSeparateProcesses
```
Comment (jasny) [inline tests/ResponseHeadersTest.php]: Don't run tests in separate process.

File: tests/ResponseHeadersTest.php [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    /**
+     * @var ResponseHeaders
+     */
+    protected $headers;
+
+    public function setUp()
+    {
+        $this->headers = new ResponseHeaders();
+    }
+
```
Comment (jasny) [inline tests/ResponseHeadersTest.php:25]: Unset `$_GLOBALS['headersList']` and `$_GLOBALS['responseCode']` after each test.

**Review by jasny — CHANGES_REQUESTED** [review]

File: tests/ResponseHeadersTest.php (hunk 11 above)
Comment (poratuk) [reply tests/ResponseHeadersTest.php:25]: https://phpunit.de/manual/current/en/fixtures.html#fixtures.more-setup-than-teardown
By default, PHPUnit runs your tests in a way where changes to global and super-global variables ($GLOBALS, $_ENV, $_POST, $_GET, $_COOKIE, $_SERVER, $_FILES, $_REQUEST) do not affect other tests. Optionally, this isolation can be extended to static attributes of classes.

**Pull request merged**